pkg/controller/endpoint/endpoints_controller.go

```diff
@@ -415,6 +415,11 @@ func (e *EndpointController) syncService(key string) error {
 				})
 				readyEps++
 			} else {
+				if (pod.Spec.RestartPolicy == v1.RestartPolicyNever && pod.Status.Phase == v1.PodFailed) ||
```

Comment:
Can you extract those conditions to a function, like shouldPodBeInEndpoints (better name would be awesome)?